### PR TITLE
Authentication system

### PR DESCRIPTION
**Pull request by nithinbekal**
This adds a simple email/password authentication to the app.

### CODE REVIEW

File: web/controllers/session_controller.ex [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def create(conn, %{"session" => session_params}) do
+    case ElixirBlog.Session.login(session_params) do
+      {:ok, user} ->
+        conn
+        |> put_session(:current_user, user.id)
+        |> put_flash(:info, "Logged in!")
+        |> redirect(to: "/")
+      {:error} ->
```
Comment (chrismccord) [inline web/controllers/session_controller.ex]: if you aren't returning an error reason, no need for a tuple here. Just `:error` is fine

File: web/models/registration.ex [hunk 2]
```diff
@@ -0,0 +1,16 @@
+defmodule ElixirBlog.Registration do
+  import Ecto.Changeset, only: [put_change: 3]
+
+  alias ElixirBlog.Repo
+
+  def create(changeset) do
```
Comment (chrismccord) [inline web/models/registration.ex]: Not necessary, but I like to pass the Repo in so we don't couple to only one. You can default it for the best of both worlds, i.e.:

``` elixir
def create(changeset, repo \\ Repo) do
  changeset
  |> put_change(:crypted_password, hashed_password(changeset.params["password"]))
  |> repo.insert()
end
```

File: web/models/session.ex [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  def login(params) do
+    user = find_user_by_email(params["email"])
+    case authenticate(user, params["password"]) do
+      true -> {:ok, user}
+      _    -> {:error}
+    end
+  end
+
+  def current_user(conn) do
+    id = Plug.Conn.get_session(conn, :current_user) || 0
```
Comment (chrismccord) [inline web/models/session.ex]: I would handle the no session case other than `0`. Don't hit the Repo if you don't need to

File: web/router.ex [hunk 4]
```diff
@@ -17,6 +17,13 @@ defmodule ElixirBlog.Router do
     get "/", PostController, :index
 
     resources "/posts", PostController
+
+    get "/register", RegistrationController, :new
+    post "/register", RegistrationController, :create
+
+    get "/login", SessionController, :new
+    post "/login", SessionController, :create
+    get "/logout", SessionController, :delete
```
Comment (chrismccord) [inline web/router.ex]: Preferable logout should not be a get request. Use `link "logout", to: session_path(@conn, :delete), method: "delete"`

File: web/controllers/session_controller.ex (hunk 1 above)
Comment (nithinbekal) [inline web/controllers/session_controller.ex]: Ah, that makes sense. I'll change that.

File: web/models/registration.ex (hunk 2 above)
Comment (nithinbekal) [inline web/models/registration.ex]: That seems better. Thanks!

Btw, should a module like this be in web/models or should it be in lib? I wasn't entirely sure about that.

File: web/models/session.ex (hunk 3 above)
Comment (nithinbekal) [inline web/models/session.ex]: Oh! I hadn't thought of that. Thanks!

File: web/router.ex (hunk 4 above)
Comment (nithinbekal) [inline web/router.ex]: Thanks for pointing that out! I'll change that.

**Comment by nithinbekal** [comment]
@chrismccord Hey Chris. Thank you so much for taking the time to review this code. And thank you for all your work on Phoenix - I'm really enjoying playing with this framework. :)

**Pull request closed without merging**